WIP: Limit offset time zones to minutes precision

justingrant opened the pull request:
This is a work-in-progress PR that fixes #2593 and fixes #2584. I'm sharing now to get early feedback, and can make changes in response to that feedback. 

The overall approach of the PR is to split everything (productions, AOs, etc.) into a "nanoseconds" and a "minutes" variant, and then use each in the appropriate place:
* Minutes are used for time zone identifiers and for ZDT/Instant `toString`/`toJSON`
* Nanoseconds are used for ZDT's `offset` property (both parsing and output), and for parsing of Instant and the RFC3339 portion of ZDT strings

Details:

* Move offset string grammar into abstractops.html, remove it from mainadditions.html
* Split the |UTCOffset| production into `|UTCOffsetMinutesPrecision|` and `|UTCOffsetNanosecondPrecision|`. 
* Split `ParseTimeZoneOffsetString` into `ParseTimeZoneOffsetMinutes` and `ParseTimeZoneOffsetNanoseconds`.
* Split `FormatTimeZoneOffsetString` into `FormatTimeZoneOffsetMinutes` and `FormatTimeZoneOffsetNanoseconds`.
* Rename `FormatISOTimeZoneOffsetString` to `FormatTimeZoneOffsetRoundToMinutes`.
* Add `IsTimeZoneOffsetStringNanosecondsPrecision`.
* Add `|NormalizedUTCOffset|` and use it in `SystemTimeZoneIdentifier` to fix #2584.
* Add a note to ISO grammar about how we extend the standard to allow 9 digits for instant string offsets.

This is a quick first draft so at this point I'm mostly interested in high-level feedback. Questions: 

* Is the overall approach OK?  Anything major or structural I should change?
* Should we rename `IsTimeZoneOffsetString` to `isTimeZoneOffsetMinutesPrecision` or something like that? If yes, then is it OK to just put an editor's note that it should be renamed throughout 262, instead of dragging hundreds of lines of 262 text into the Temporal spec just for the rename?
* Should `ParseTimeZoneOffsetMinutes` and `FormatTimeZoneOffsetMinutes` return and accept, respectively, minutes or nanoseconds? Personally I find it clearer to use minutes, but that's more spec churn. Either way is OK with me.
* Are the new productions correct?  I'm not a BNF expert so expect errors there.

gibson042 commented on spec/mainadditions.html:
```diff
@@ -271,6 +271,247 @@ <h1>AvailableNamedTimeZoneIdentifiers ( ): a List of Time Zone Identifier Record
+
+      <emu-alg>
+        1. If the implementation only supports the UTC time zone, return *"UTC"*.
+        1. Let _systemTimeZoneString_ be the String representing the host environment's current time zone, either a primary time zone identifier or <del>an offset time zone identifier</del><ins>a |NormalizedUTCOffset|</ins>.
```
```suggestion
        1. Let _systemTimeZoneString_ be the String representing the host environment's current time zone, either a primary time zone identifier or <del>an offset time zone identifier</del><ins>a String _s_ for which ParseText(StringToCodePoints(_s_), |NormalizedUTCOffset|) returns a Parse Node</ins>.
```

justingrant replied:
I wrote this by copying a bunch of examples in 262 text using "is a |Foo|", which does seem clearer than using ParseText. Examples: 

* Each [[Site]] value is a Parse Node that is a |TemplateLiteral|
* String that when processed is a |FunctionBody|
* Assert: _lhs_ is a |LeftHandSideExpression|.
* Assert: _d_ is a |FunctionDeclaration|.

Is there a reason to use the longer text instead of the simpler one?

justingrant replied:
Or is the difference that those 262 examples are Code Points and not Strings, so you need the extra wording?  If so, should we create a new AO e.g. `StringMatchesProduction` or something like that to make that case less wordy?

justingrant replied:
After looking into this more, it's the String vs. Code Points thing. Suggestion accepted!

gibson042 replied:
See also https://github.com/tc39/ecma262/pull/3099 , which if accepted will allow for some minor improvement here.

justingrant replied:
That 262 PR would simplify Temporal, good idea. In the meantime I'll just use your suggestion. Only problem: it causes a spec lint error: `could not find a preceding declaration for "s"`. Any idea what is alternate text that won't cause that error?

gibson042 replied:
Yep: https://tc39.es/ecmarkup/#variable-declarations
```
        1. <ins>[declared="s"]</ins> Let _systemTimeZoneString_ be the String representing the host environment's current time zone, either a primary time zone identifier or <del>an offset time zone identifier</del><ins>a String _s_ for which ParseText(StringToCodePoints(_s_), |NormalizedUTCOffset|) returns a Parse Node</ins>.
```

justingrant replied:
None of those prevent the lint error. 😢 

How about I just leave it as "is a |NormalizedUTCOffset|" and hope that your 262 PR gets merged before the July plenary when we present this PR?

gibson042 replied:
Ok, but please also open an issue in https://github.com/tc39/ecmarkup. I suspect it is an \<ins>/\<del> problem.

justingrant replied:
👍 https://github.com/tc39/ecmarkup/issues/528

justingrant replied:
There's a workaround, which is to put the `[declared ...]` outside the `<ins>`. New text:

```
1. [declared="s"] Let _systemTimeZoneString_ be the String representing the host environment's current time zone, either a primary time zone identifier or an offset time zone identifier<ins> String _s_ for which ParseText(StringToCodePoints(_s_), |NormalizedUTCOffset|) returns a Parse Node</ins>.
```

gibson042 commented on spec/mainadditions.html:
```diff
@@ -271,6 +271,247 @@ <h1>AvailableNamedTimeZoneIdentifiers ( ): a List of Time Zone Identifier Record
+      <ins class="block">
+      <emu-clause id="sec-interprettimezoneoffsetparseresult" type="abstract operation">
+        <h1>
+          InterpretTimeZoneOffsetParseResult (
```
I'd also be happier to see this integrated into a single Record-returning ParseTimeZoneOffsetString.

gibson042 commented on spec/timezone.html:
```diff
@@ -489,21 +500,31 @@ <h1>FormatTimeZoneOffsetString ( _offsetNanoseconds_ )</h1>
+    </emu-clause>
+
+    <emu-clause id="sec-temporal-formatisotimezoneoffsetstring" type="abstract operation">
+      <h1>FormatTimeZoneOffsetRoundToMinutes ( _offsetNanoseconds_ )</h1>
```
Suggestion: combine all operations that format a UTC offset (including particularly FormatTimeZoneOffsetMinutes and FormatTimeZoneOffsetRoundToMinutes) into a single operation.

justingrant replied:
Makes sense, can do.

justingrant replied:
I ended up reverting changes to formatting. So there's still two formatting methods: FormatTimeZoneOffsetString and FormatISOTimeZoneOffsetString. The only change I made was to make FormatISOTimeZoneOffsetString call FormatTimeZoneOffsetString to do the actual formatting. Otherwise no changes. Holler if this is a problem.

gibson042 reviewed (commented):
I got a little in, but honestly am feeling overwhelmed by the plethora of abstract operations. For me at least, I think the splits hinder comprehension.

justingrant commented on spec/abstractops.html:
```diff
@@ -1098,14 +1099,25 @@ <h1>ISO 8601 grammar</h1>
       TimeZoneUTCOffsetName[Extended] :
           Sign Hour[+Padded]
           Sign Hour[+Padded] TimeSeparator[?Extended] MinuteSecond
-          Sign Hour[+Padded] TimeSeparator[?Extended] MinuteSecond TimeSeparator[?Extended] MinuteSecond Fraction?
```
Refactor looks good, but after trying exactly the naming you recommended (where nano precision is just |UTCOffset|) and trying the current PR text's naming (where both the nano and the minute version have the precision in the name) is that I found it easier to reason about which call sites use which variant of the offset syntax when I didn't have to keep remembering which was the nano version and which was the minutes version.  

Also, if we did call one of them just plain |UTCOffset|, then I'd probably suggest it be the minutes version, because that's the one supported by external standards and also the one actually supported by current shipping implementations.  The nano version, being the new kid on the block, seems like it deserves a special name to let folks know that it's a standards deviation. 

What do you think?

gibson042 replied:
I guess I'd want to see a side-by-side comparison, but my gut reaction is that the name qualifiers matter less than the grammar structure.

justingrant replied:
OK cool. I'll plan to use the structure you suggested but probably will use a different name than |UTCOffset|, for clarity.

justingrant replied:
OK, refactor is pushed. Want to take a look and let me know what you think?

justingrant commented on spec/abstractops.html:
```diff
@@ -1409,8 +1421,8 @@ <h1>
       1. Else,
-        1. If _parseResult_ contains a |UTCOffset| Parse Node, then
-          1. Let _offset_ be the source text matched by the |UTCOffset| Parse Node contained within _parseResult_.
+        1. If _parseResult_ contains a |UTCOffsetMinutePrecision| Parse Node, then
+          1. Let _offset_ be the source text matched by the |UTCOffsetMinutePrecision| Parse Node contained within _parseResult_.
```
Good catch! This should be |UTCOffsetNanosecondPrecision| because it's an RFC3339 string, not a time zone identifier. Fixing in latest commit.

justingrant commented on spec/abstractops.html:
```diff
@@ -622,8 +622,8 @@ <h1>ToRelativeTemporalObject ( _options_ )</h1>
       1. If _offsetBehaviour_ is ~option~, then
-        1. If IsTimeZoneOffsetString(_offsetString_) is *false*, throw a *RangeError* exception.
-        1. Let _offsetNs_ be ParseTimeZoneOffsetString(_offsetString_).
+        1. If IsTimeZoneOffsetStringNanosecondPrecision(_offsetString_) is *false*, throw a *RangeError* exception.
+        1. Let _offsetNs_ be ParseTimeZoneOffsetStringNanoseconds(_offsetString_).
```
I like the idea of a single parsing AO and a single formatting AO, and considered doing that initially. 

Where I got tripped up was what to do about IsTimeZoneOffsetString[NanosecondPrecision]. That's used in ~20 places in spec text, including a bunch inside 262 that'd have to be dragged into Temporal if we removed those AOs.

Do you think we should retain those Is* AOs?  If yes, their trivial implementation could use the (single) parsing AO.

Somewhat related: is there any way to avoid the DRY noise of the following text being repeated 15+ times?
```
1. If _offsetRecord_ is a List of errors, throw a *RangeError* exception.
  1. If _offsetRecord_.[[Precision]] is not ~hour~ or ~minute~, throw a *RangeError* exception.
```

If we're already building a custom parsing AO for offsets, could this just be one line?

What I was thinking about would be a record with these fields:
* [[Nanoseconds]] - integer or `~empty~` if not matched
* [[Minutes]] - integer or `~empty~` if not matched

I'm not sure there's a lot of value supporting other precisions (hour, second, etc.) if nothing in the spec actually uses those.

Finally, does having standalone AOs make it easier for implementers to create optimized paths for each format?  I don't have an opinion about this, just wondering.

gibson042 replied:
> Do you think we should retain those Is* AOs? If yes, their trivial implementation could use the (single) parsing AO.

Sure, as long as things are structured for their later removal.

> 
> Somewhat related: is there any way to avoid the DRY noise of the following text being repeated 15+ times?
> 
> ```
> 1. If _offsetRecord_ is a List of errors, throw a *RangeError* exception.
>   1. If _offsetRecord_.[[Precision]] is not ~hour~ or ~minute~, throw a *RangeError* exception.
> ```
> 
> If we're already building a custom parsing AO for offsets, could this just be one line?
> 
> What I was thinking about would be a record with these fields:
> 
> * [[Nanoseconds]] - integer or `~empty~` if not matched
> * [[Minutes]] - integer or `~empty~` if not matched
> 
> I'm not sure there's a lot of value supporting other precisions (hour, second, etc.) if nothing in the spec actually uses those.

I'd pretty strongly prefer to specify all numeric data in terms of the same units (i.e., nanoseconds). Having a mixture seems to invite confusion and possibly implementation bugs, and going all the way to a Duration Record is too far for the same reason you suggest here. So the shape wouldn't be `{ [[Minutes]]: integer | ~empty~, [[Nanoseconds]]: integer | ~empty~ }`, but could be e.g. `{ [[Nanoseconds]]: integer | ~empty~, [[HasSubMinutePrecision]]: Boolean }}`.

> Finally, does having standalone AOs make it easier for implementers to create optimized paths for each format? I don't have an opinion about this, just wondering.

I don't think there's a direct connection from one to the other in most cases; for me it's more of a judgment call about the cognitive burden of spec comprehension.

justingrant replied:
> but could be e.g. `{ [[Nanoseconds]]: integer | ~empty~, [[HasSubMinutePrecision]]: Boolean }}`

👍

justingrant replied:
I just pushed an update that changes parsing to return records. It worked better to have two different parsing methods: one for time zone identifiers and one for RFC 3339 string offsets.  Here's what I've got so far. Let me know what you think:

* Split `ParseTimeZoneOffsetString` into two new AOs for parsing ISO string offsets and offset time zone identifiers, respectively. Both return Records instead of primtives:
  * `ParseUTCOffsetString` returns { [[OffsetNanoseconds]: an integer, [[HasSubMinutePrecision]]: a Boolean }
  * `ParseTimeZoneIdentifier` returns { [[Name]]: a String or `~empty~`, [[OffsetMinutes]]: an integer or `~empty~`}
  * If the input string isn't valid, both new AOs return `~unmatched~` instead of a Record.

justingrant commented:
V2 is available. It's a refactor based on @gibson042's advice around parsing and grammar. 

Let me know if you see problems with this version. If not then I'll move onto Test262.

justingrant commented:
Also, I updated the OP so you can see a summary of the latest design.

justingrant commented:
@gibson042 @ptomato PR is now ready for review. I split it up into one editorial commit that refactors offset handling, and two normative commits: 
* A small one that ensures that SystemTimeZoneIdentifier only returns offset strings in normal form
* A big one that makes the nanoseconds=>minutes change for offset time zones

I'll work on Test262 next.

gibson042 commented on spec/timezone.html:
```diff
@@ -464,10 +472,21 @@ <h1>
+        <dt>description</dt>
+        <dd>
+          It formats a time zone offset, in nanoseconds, into a UTC offset string.
+          If _offsetNanoseconds_ represents an integer number of minutes, then the output will be in ±HH:MM format.
+          Otherwise, the output will be in ±HH:MM:SS[.S[S[S[S[S[S[S[S[S]]]]]]]]] format.
```
```suggestion
          If _offsetNanoseconds_ represents an integer number of minutes, then the output will be formatted like ±HH:MM.
          Otherwise, the output will be formatted like ±HH:MM:SS or (if the offset does not evenly divide into seconds) ±HH:MM:SS.fff… where the "fff" part is a sequence of at least 1 and at most 9 fractional seconds digits with no trailing zeroes.
```

gibson042 commented on spec/timezone.html:
```diff
@@ -489,21 +508,16 @@ <h1>FormatTimeZoneOffsetString ( _offsetNanoseconds_ )</h1>
 
-    <emu-clause id="sec-temporal-formatisotimezoneoffsetstring" aoid="FormatISOTimeZoneOffsetString">
+    <emu-clause id="sec-temporal-formatisotimezoneoffsetstring" type="abstract operation">
       <h1>FormatISOTimeZoneOffsetString ( _offsetNanoseconds_ )</h1>
```
If there are going to be two operations, I think this one should be renamed to something like "FormatTimeZoneOffsetStringISO" so there is a unique prefix corresponding with the pair.

justingrant replied:
What about `FormatDateTimeUTCOffset`?  This aligns with the new production |DateTimeUTCOffset|. I can also rename the parsing version of it to be `ParseDateTimeUTCOffset` for consistency and clarity.

justingrant replied:
Or maybe `FormatDateTimeUTCOffsetRounded` to make super-clear that there will be rounding going on?

justingrant replied:
I just pushed `FormatDateTimeUTCOffsetRounded` in the latest commit, let me know if you'd like to see something different.

gibson042 reviewed (changes requested):
There's a lot of good work here, but I think a little more is still warranted.

justingrant commented on spec/timezone.html:
```diff
@@ -810,5 +810,34 @@ <h1>
+      <h1>
+        ParseTimeZoneIdentifier (
+          _identifier_: a String
+        ): a Record containing [[Name]] and [[OffsetMinutes]] fields, or ~unmatched~
```
Thanks for your thorough review!  I'll commit the rest of your suggestions, but wanted to dig in on this one a bit more.

The switch to [[OffsetMinutes]] was intended to align with the main goal of this PR: to enable a Temporal.TimeZone to be stored in <=16 bits.  Unlike some optimizations that some implementations will use and others ignore, I assume that all native implementations will store minutes not nanoseconds.

Given that implementations will be doing the conversion in their own implementation, I thought it would be better to specify the conversion math so that the spec would align more closely with how it will be implemented. In particular I was sensitive to feedback from Shu that Temporal implementations were so large and complex that the only scalable way to code-review them is to ensure that they matched the spec as closely as possible.

So I actually think that specifying the slot as minutes will lead to _less_ buggy implementations, not more.

Also, I liked how using [[OffsetMinutes]] prevents readers/reviewers/implementers from having to worry about accidentally introducing out-of-range data via nanoseconds sneaking in. I generally like it when validation is baked into the data type and the name, as opposed to having to remember and reason about implicit constraints that are not recorded in the type or the name, like "this may be a nanoseconds value that represents an integer number of minutes".

What do you think?

Another kinda-compromise option (inspired by V1 of this PR): of the 6 places where the minutes=>nanoseconds calculation is done in the current PR text, all but one are formatting minutes into an offset string. There's also only one place (GetOffsetStringFor) where a nanoseconds offset is formatted; the rest are minutes only. So what if we did this something like this?
* Add a `FormatOffsetTimeZoneIdentifier` AO. This would limit minutes=>nanoseconds calculation to only two places: the formatting AO and a single line in `TimeZone.prototype.getPossibleInstantsFor`. This would also provide symmetry with the reverse operation ParseTimeZoneIdentifier.
* Remove FormatTimeZoneOffsetString and inline nanoseconds formatting into `GetOffsetStringFor`. Something like this: (rough draft; needs editing)
```
1. Let _offsetNanoseconds_ be ? GetOffsetNanosecondsFor(_timeZone_, _instant_).
1. Let _subMinuteNanoseconds_ be _offsetNanoseconds_ modulo (6 × 10<sup>10</sup>).
1. Let _minutes_ be  (_offsetNanoseconds_ - _subMinuteNanoseconds_) / (6 × 10<sup>10</sup>).
1. Let _offsetString_ be FormatOffsetTimeZoneIdentifier(_minutes_).
1. If _subMinuteNanoseconds_ = 0, return _offsetSring_.
1. <text to calculate seconds and optional sub-seconds, then concatenate to _offsetString_.
1. Return _offsetString_
```

justingrant replied:
> of the 6 places where the minutes=>nanoseconds calculation is done in the current PR text, all but one are formatting minutes into an offset string

Actually, I missed three: `LocalTime`, `UTC`, and `TimeZoneString` in existing 262. What was nice was that this error was pretty obvious because the field names were wrong. ;-)

So switching the formatting method to minutes would only clean up 5 of 9 uses of  `× 10<sup>6</sup`, not 5 of 6 like I thought before. Still might be useful DRY though.

gibson042 replied:
> So what if we did this something like this?
> 
> * Add a `FormatOffsetTimeZoneIdentifier` AO. This would limit minutes=>nanoseconds calculation to only two places: the formatting AO and a single line in `TimeZone.prototype.getPossibleInstantsFor`. This would also provide symmetry with the reverse operation `ParseTimeZoneIdentifier`.
> * Remove `FormatTimeZoneOffsetString` and inline sub-minute formatting into `GetOffsetStringFor`.

Yes, that seems pretty good!

justingrant replied:
OK, I made the changes above. Take a look at the new text and let me know what you think.

justingrant commented on spec/mainadditions.html:
```diff
@@ -271,6 +271,266 @@ <h1>AvailableNamedTimeZoneIdentifiers ( ): a List of Time Zone Identifier Record
+        <dt>description</dt>
+        <dd>The return value indicates whether _offsetString_ conforms to the grammar given by <del>|UTCOffset|</del><ins>|TimeZoneUTCOffsetName|</ins>.</dd>
+      </dl>
+      <emu-note type="editor">
+        This abstract operation was renamed from IsTimeZoneOffsetString.
+        I couldn't figure out how to get `ins` and `del` to work with structured headers.
+      </emu-note>
```
Cool, thanks for that. I added a reference to the issue in the PR spec text as a reminder to change it later after ecmarkup is fixed.

justingrant commented on spec/mainadditions.html:
```diff
@@ -271,6 +271,266 @@ <h1>AvailableNamedTimeZoneIdentifiers ( ): a List of Time Zone Identifier Record
+      <emu-alg>
+        1. Let _systemTimeZoneIdentifier_ be SystemTimeZoneIdentifier().
+        1. <ins>Let _parseResult_ be ParseTimeZoneIdentifier(_systemTimeZoneIdentifier_).</ins>
+        1. If <del>IsTimeZoneOffsetString(_systemTimeZoneIdentifier_) is *true*</del><ins>>_parseResult_.[[OffsetNanoseconds]] is not ~empty~</ins>, then
```
Search-n-replace gone awry. Thanks for catching this!

gibson042 commented on spec/mainadditions.html:
```diff
@@ -271,6 +271,266 @@ <h1>AvailableNamedTimeZoneIdentifiers ( ): a List of Time Zone Identifier Record
+        <dt>description</dt>
+        <dd>The return value indicates whether _offsetString_ conforms to the grammar given by <del>|UTCOffset|</del><ins>|TimeZoneUTCOffsetName|</ins>.</dd>
+      </dl>
+      <emu-note type="editor">
+        This abstract operation was renamed from IsTimeZoneOffsetString.
+        After https://github.com/tc39/ecmarkup/issues/530 is fixed, this note can be replaced with `ins` and `del` in the AO name above.
+      </emu-note>
```
Fixed now in ecmarkup v17.0.1!
https://github.com/tc39/ecmarkup/pull/533#issuecomment-1596272123

justingrant replied:
I don't think this is actually fixed, at least I couldn't figure out how to make it work: https://github.com/tc39/ecmarkup/pull/533#issuecomment-1596453212

gibson042 commented on spec/timezone.html:
```diff
@@ -112,7 +116,7 @@ <h1>Temporal.TimeZone.prototype.getOffsetNanosecondsFor ( _instant_ )</h1>
         1. Let _timeZone_ be the *this* value.
         1. Perform ? RequireInternalSlot(_timeZone_, [[InitializedTemporalTimeZone]]).
         1. Set _instant_ to ? ToTemporalInstant(_instant_).
-        1. If _timeZone_.[[OffsetNanoseconds]] is not ~empty~, return 𝔽(_timeZone_.[[OffsetNanoseconds]]).
+        1. If _timeZone_.[[OffsetMinutes]] is not ~empty~, return 𝔽(_timeZone_.[[OffsetMinutes]]).
```
Looks like this change is incomplete; `getOffsetNanosecondsFor` should not return a number of _minutes_.

gibson042 commented on spec/timezone.html:
```diff
@@ -277,11 +281,11 @@ <h1>Properties of Temporal.TimeZone Instances</h1>
-              An integer for nanoseconds representing the constant offset of this time zone relative to UTC, or ~empty~ if the instance represents a named time zone.
-              If not ~empty~, this value must be greater than &minus;8.64 &times; 10<sup>13</sup> nanoseconds (&minus;24 hours) and smaller than &plus;8.64 &times; 10<sup>13</sup> nanoseconds (&plus;24 hours).
+              An integer for minutes representing the constant offset of this time zone relative to UTC, or ~empty~ if the instance represents a named time zone.
+              If not ~empty~, this value must be greater than &minus;1440 minutes (&minus;24 hours) and smaller than &plus;1440 minutes (&plus;24 hours).
```
```suggestion
              If not ~empty~, this value must be in the interval from -1440 (exclusive) to +1440 (exclusive) (i.e., strictly less than 24 hours in magnitude).
```
(refer to [interval](https://tc39.es/ecma262/multipage/notational-conventions.html#interval) for examples of "interval" variations and [Expanded Years](https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-expanded-years) for examples of signed integers)

justingrant replied:
Cool. To make ecmarkup happy, I had to remove the `+` from the positive number. Final text: 

```
If not ~empty~, this value must be in the interval from -1440 (exclusive) to 1440 (exclusive) (i.e., strictly less than 24 hours in magnitude).
```

justingrant commented on spec/timezone.html:
```diff
@@ -277,11 +281,11 @@ <h1>Properties of Temporal.TimeZone Instances</h1>
           <tr>
             <td>
-              [[OffsetNanoseconds]]
+              [[OffsetMinutes]]
```
Per our discussion in https://github.com/tc39/proposal-temporal/pull/2607/files#r1233183572, I left this as [[OffsetMinutes]]. Let me know if this is a problem!

justingrant commented on spec/timezone.html:
```diff
@@ -32,7 +32,11 @@ <h1>Temporal.TimeZone ( _identifier_ )</h1>
+        1. Let _parseResult_ be ParseTimeZoneIdentifier(_identifier_).
+        1. If _parseResult_ is ~unmatched~, throw a *RangeError* exception.
+        1. If _parseResult_.[[OffsetMinutes]] is not ~empty~, then
+          1. Set _identifier_ to FormatTimeZoneOffsetString(_parseResult_.[[OffsetMinutes]] × (6 × 10<sup>10</sup>)).
```
Formatting-argument math is now avoided by making the formatting AO accept only minutes. There's still some math in other places where nanoseconds are needed, but this formatting AO change removed about half of the math instances.

ptomato replied:
I think I agree with Richard that I'd rather keep the internal representation of UTC offsets as nanoseconds the whole way through. I get that implementors might store it as minutes, but they can make that optimization on their own. Keeping one representation throughout the spec seems better for comprehension.

justingrant replied:
Could we continue this discussion in https://github.com/tc39/proposal-temporal/pull/2607#discussion_r1233183572?

I listed a few reasons there why I think the minutes vs. nanoseconds split will make implementation bugs less likely, and I'd be interested whether you think those reasons outweigh the extra complexity of the split.

I'll resolve this comment now so we can focus discussion over there.

justingrant commented on spec/mainadditions.html:
```diff
@@ -271,6 +271,266 @@ <h1>AvailableNamedTimeZoneIdentifiers ( ): a List of Time Zone Identifier Record
+      </emu-note>
+      <emu-alg>
+        1. Let _parseResult_ be ParseText(StringToCodePoints(_offsetString_), <del>|UTCOffset|</del><ins>|UTCOffsetSubMinutePrecision|</ins>).
+        1. <ins>If _parseResult_ is a List of errors, return ~unmatched~.</ins>
```
BTW, the same was true for `ParseTimeZoneIdentifier`, so I added throwing to that AO too.

justingrant commented on spec/mainadditions.html:
```diff
@@ -271,6 +271,258 @@ <h1>AvailableNamedTimeZoneIdentifiers ( ): a List of Time Zone Identifier Record
+      <emu-alg>
+        1. If the implementation only supports the UTC time zone, return *"UTC"*.
+        1. [declared="s"] Let _systemTimeZoneString_ be the String representing the host environment's current time zone, either: a primary time zone identifier; or an offset time zone identifier<ins> String _s_ for which ParseText(StringToCodePoints(_s_), |NormalizedUTCOffset|) returns a Parse Node</ins>.
+        1. <del>If IsTimeZoneOffsetString(_systemTimeZoneString_) is *true*, return _systemTimeZoneString_.</del>
```
This weird line was left over from 262 editors' comments to move the assertions into the previous line. But after assertions were removed, this line became unnecessary.

justingrant commented on spec/mainadditions.html:
```diff
@@ -271,6 +271,269 @@ <h1>AvailableNamedTimeZoneIdentifiers ( ): a List of Time Zone Identifier Record
+          1. <del>Let _absOffset_ be -_offset_.</del>
+        1. <del>Let _offsetMin_ be ToZeroPaddedDecimalString(ℝ(MinFromTime(_absOffset_)), 2).</del>
+        1. <del>Let _offsetHour_ be ToZeroPaddedDecimalString(ℝ(HourFromTime(_absOffset_)), 2).</del>
+        1. <ins>Let _offsetString_ be FormatOffsetTimeZoneIdentifier(_offsetNs_, ~legacy~).</ins>
```
This line replaces the deleted text above, but I'm also happy to revert this (and to eliminate the `~legacy~` style of `FormatOffsetTimeZoneIdentifier` if we don't want to change anything involved with `Date`. Let me know which is best.

ptomato reviewed (approved):
Seems good. I'd really prefer to keep offset nanoseconds as the quantity used in the spec text, though, and not use offset minutes.

I would suggest moving the editorial commit into a separate PR so that this one is smaller for TC39 delegates to review.

justingrant commented:
> Seems good. I'd really prefer to keep offset nanoseconds as the quantity used in the spec text, though, and not use offset minutes.

Could you take a look at https://github.com/tc39/proposal-temporal/pull/2607#discussion_r1233183572?  Richard and I came to at least a partial consensus that I think addressed many of his concerns about this.  And I think there are good reasons (noted in the thread linked above) to split minutes (for identifiers) from nanoseconds (for date/time offsets) to make it clearer for implementers that these are different values that will always require some sort of conversion and/or validation step between them. I'd be interested in hearing your take too!

> I would suggest moving the editorial commit into a separate PR so that this one is smaller for TC39 delegates to review.

OK, I'll plan to do that.

Thanks!

justingrant commented:
Tests are ready at https://github.com/tc39/test262/pull/3862.

> Seems good. I'd really prefer to keep offset nanoseconds as the quantity used in the spec text, though, and not use offset minutes.
> 
> I would suggest moving the editorial commit into a separate PR so that this one is smaller for TC39 delegates to review.

To help make sure @ptomato's vacation is worry-free, I'm planning to do these changes next week:
1. Split out the first commit in this PR into a separate editorial PR
2. Limit this PR to only the normative changes, stacked on (1)
3. Split out the refactoring of the TImeZone slot from OffsetNanoseconds to OffsetMinutes into a separate editorial PR, stacked on (2).  This has an additional advantage of making the normative PR easier to review.

We can have the discussion about (3) after Philip is back.

justingrant commented:
@gibson042 - you still have a "changes requested" status on your review. If you have a minute, could you verify that the changes you wanted to see are completed?  Note that this PR will is now much smaller to review because the editorial commits were moved to other PRs.

gibson042 commented on spec/timezone.html:
```diff
@@ -657,11 +647,29 @@ <h1>
+        1. If _nanoseconds_ &ne; 0, then
+          1. Let _fraction_ be ToZeroPaddedDecimalString(_nanoseconds_, 9).
+          1. Set _fraction_ to the longest possible substring of _fraction_ starting at position 0 and not ending with the code unit 0x0030 (DIGIT ZERO).
+          1. Let _post_ be the string-concatenation of the code unit 0x003A (COLON), _s_, the code unit 0x002E (FULL STOP), and _fraction_.
+        1. Else if seconds &ne; 0, then
+          1. Let _post_ be the string-concatenation of the code unit 0x003A (COLON) and _s_.
+        1. Else,
+          1. Let _post_ be the empty String.
+        1. Return the string-concatenation of _offsetString_ and _post_.
```
Isn't the final Else branch impossible to reach?
```suggestion
          1. Let _suffix_ be the string-concatenation of the code unit 0x002E (FULL STOP) and _fraction_.
        1. Else,
          1. Assert: _seconds_ &ne; 0.
          1. Let _suffix_ be *""*.
        1. Return the string-concatenation of _offsetString_, the code unit 0x003A (COLON), _s_, and _suffix_.
```

justingrant replied:
Good catch, yes it's now unreachable. Will fix.

justingrant replied:
Since we're refactoring this AO anyways, I made a few more editorial changes that IMO better communicate the cascade of early returns if seconds are zero, and if not then if nanos are zero, and finally to fall through to the nano case.

```
1. Let _offsetNanoseconds_ be ? GetOffsetNanosecondsFor(_timeZone_, _instant_).
1. Let _offsetMinutes_ be truncate(_offsetNanoseconds_ / (6 × 10<sup>10</sup>)).
1. Let _offsetString_ be FormatOffsetTimeZoneIdentifier(_offsetMinutes_ × (6 × 10<sup>10</sup>), ~separate~).
1. Let _subMinuteNanoseconds_ be abs(_offsetNanoseconds_) modulo (6 × 10<sup>10</sup>).
1. If _subMinuteNanoseconds_ = 0, then
  1. Return _offsetString_.
1. If _offsetMinutes_ = 0 and _offsetNanoseconds_ &lt; 0, set _offsetString_ to *"-00:00"*.
1. Let _seconds_ be floor(_subMinuteNanoseconds_ / 10<sup>9</sup>) modulo 60.
1. Let _ss_ be ToZeroPaddedDecimalString(_seconds_, 2).
1. Let _nanoseconds_ be _subMinuteNanoseconds_ modulo 10<sup>9</sup>.
1. If _nanoseconds_ = 0, then
  1. Return the string-concatenation of _offsetString_, the code unit 0x003A (COLON), and _ss_.
1. Let _fraction_ be ToZeroPaddedDecimalString(_nanoseconds_, 9).
1. Set _fraction_ to the longest prefix of _fraction_ ending with a code unit other than 0x0030 (DIGIT ZERO).
1. Return the string-concatenation of _offsetString_, the code unit 0x003A (COLON), _ss_, the code unit 0x002E (FULL STOP), and _fraction_.
```

gibson042 commented on spec/mainadditions.html:
```diff
@@ -228,6 +227,15 @@ <h1>Time Zone Identifiers</h1>
       </p>
+      <ins class="block">
+      <p>
+        Time zone identifiers are compared using ASCII-case-insensitive comparisons, and are accepted as input in any variation of letter case.
```
```suggestion
        Named time zone identifiers are compared using ASCII-case-insensitive comparisons, and are accepted as input in any variation of letter case.
```

justingrant replied:
We don't actually have a definition for "named time zone" or "named time zone identifier". We only have "available named time zone", "offset time zone", and "time zone identifier". In lieu of introducing a new definition, I opted to use the more general "time zone identifiers" here. I didn't want to use "available named time zone identifier" here because even custom time zone IDs should be compared case-insensitively. And offset time zones can't include letters, so "Time zone identifiers are compared using ASCII-case-insensitive comparisons" is correct even though it's less precise than it could be.

Anyway, the alternative is to add a new definition for "named time zone identifier" that includes both built-in IDs ("available named time zone identifiers") and custom IDs ("custom named time zone identifiers"?). But naming the latter will involve some bikeshedding and I wasn't sure it was strictly needed. 

What do you think?  My preference would be to leave this as-is.

gibson042 replied:
I find the current text misleading, because comparison logic for offset time zones is distinct from the logic for named time zones. And given a goal of resolving that without introducing any other new terms, I would suggest something like
> A time zone identifier is a _<dfn>normalized time zone identifier</dfn>_ when it is either an available named time zone identifier or can be matched by _|NormalizedUTCOffset|_. Every time zone identifier produced by an implementation must be a normalized time zone identifier, but the corresponding time zone is also identified by input that is an ASCII-case-insensitive match for its normalized identifier, and an offset time zone is additionally identified by input representing the same numeric offset in a form that can be matched by _|TimeZoneUTCOffsetName|_. For example, input of **"america/los_angeles"** and **"America/Los_angeles"** both correspond with the normalized time zone identifier **"America/Los_Angeles"**, and input of **"-08"** and **"−08"** and **"-0800"** all correspond with the normalized time zone identifier **"-08:00"**.

justingrant replied:
I still have serious PTSD from trying to get this text into ECMA-262 in our earlier PR that required 4+ rounds of feedback with the 262 editors. My sense from going through that process was that brevity was more important to them and they were willing to sacrifice details and explanation to keep things shorter. I agree your text above is clearer, but I suspect that it would just be stripped out by the editors later, so I'm loath to add it now as part of this PR.

So if we were to change this, I'd probably want to add a new term instead of adding a lot of new text.

Also, "Time zone identifiers are compared using ASCII-case-insensitive comparisons" is already in the Temporal spec; this PR just moves it to a different paragraph.

Taking these together, I think it's best to merge this text as-is, and then if we want to clarify it, to use a separate editorial PR for that purpose. Sound OK?

gibson042 commented on spec/mainadditions.html:
```diff
@@ -283,7 +291,7 @@ <h1>SystemTimeZoneIdentifier ( ): a String</h1>
       <emu-alg>
         1. If the implementation only supports the UTC time zone, return *"UTC"*.
-        1. Let _systemTimeZoneString_ be the String representing the host environment's current time zone, either: a primary time zone identifier; or an offset time zone identifier.
+        1. Let _systemTimeZoneString_ be the String representing the host environment's current time zone, either a primary time zone identifier or an offset time zone identifier<ins>, as a time zone identifier in normalized format</ins>.
```
```suggestion
        1. Let _systemTimeZoneString_ be the String representing the host environment's current time zone <ins>as a time zone identifier in normalized format</ins>, either a primary time zone identifier or an offset time zone identifier.
```

justingrant reviewed (commented):
Thanks @gibson042 for the review!  I agree with all your comments except one commented below, but we're about to present this PR in a few hours and I think all your comments are editorial, so I'm going to defer making those changes until after we present this PR to plenary. Holler if that's not OK!  FYI @ptomato just in case you have an opinion about this too.

ptomato commented:
I think it's fine to make those minor editorial changes in the coming days. Let me know when everyone is happy about it, and I'll merge https://github.com/tc39/test262/pull/3862 and follow it up by pushing a commit to update test262 and then merge this.

ptomato commented:
For the record, this normative change achieved consensus in TC39 at the plenary meeting of 2023-07-12.

justingrant commented on spec/timezone.html:
```diff
@@ -482,31 +483,20 @@ <h1>
         1. If _offsetNanoseconds_ &ge; 0, let _sign_ be the code unit 0x002B (PLUS SIGN); otherwise, let _sign_ be the code unit 0x002D (HYPHEN-MINUS).
-        1. Set _offsetNanoseconds_ to abs(_offsetNanoseconds_).
-        1. Let _hours_ be floor(_offsetNanoseconds_ / (3.6 × 10<sup>12</sup>)).
+        1. Assert: _offsetNanoseconds_ modulo (6 × 10<sup>10</sup>) = 0.
+        1. Let _offsetMinutes_ be abs(_offsetNanoseconds_ / (6 × 10<sup>10</sup>)).
```
Great suggestion. Used it.

justingrant commented on spec/timezone.html:
```diff
@@ -657,11 +647,29 @@ <h1>
+        1. Let _nanoseconds_ be _subMinuteNanoseconds_ modulo 10<sup>9</sup>.
+        1. If _nanoseconds_ &ne; 0, then
+          1. Let _fraction_ be ToZeroPaddedDecimalString(_nanoseconds_, 9).
+          1. Set _fraction_ to the longest possible substring of _fraction_ starting at position 0 and not ending with the code unit 0x0030 (DIGIT ZERO).
```
The same text was used in FormatSecondsStringPart, so I replaced both places with your suggestion.

justingrant commented:
All review comments are addressed and PR is rebased to main. 

Tests over in https://github.com/tc39/test262/pull/3862 are I think also ready to merge if we're OK merging them despite concerns in https://github.com/tc39/test262/pull/3862/files#r1259590953 that I'd prefer to defer for now.

gibson042 commented on spec/timezone.html:
```diff
@@ -657,11 +647,26 @@ <h1>
-        1. Return FormatOffsetTimeZoneIdentifier(_offsetNanoseconds_, ~separated~).
+        1. Let _offsetMinutes_ be truncate(_offsetNanoseconds_ / (6 × 10<sup>10</sup>)).
+        1. Let _offsetString_ be FormatOffsetTimeZoneIdentifier(_offsetMinutes_ × (6 × 10<sup>10</sup>), ~separate~).
+        1. Let _subMinuteNanoseconds_ be abs(_offsetNanoseconds_) modulo (6 × 10<sup>10</sup>).
```
I can't help but notice the smell of FormatOffsetTimeZoneIdentifier and GetOffsetStringFor performing some of the same arithmetic, and for that matter overlapping so much with FormatSecondsStringPart. It would be great to see reduction of FormatOffsetTimeZoneIdentifier to something like `Assert: _offsetNanoseconds_ modulo (6 × 10<sup>10</sup>) = 0. Return FormatUTCOffset(_offsetNanoseconds_, *":"*).` and GetOffsetStringFor to something like `Let _offsetNanoseconds_ be ? GetOffsetNanosecondsFor(_timeZone_, _instant_). Return FormatUTCOffset(_offsetNanoseconds_, *":"*).`, with FormatUTCOffset calculating _sign_ and returning the concatenation of it with the result of sending the decomposition of nanoseconds to a time-formatting operation (which should contain **all** of the hh:mm[:ss[.fff…]] formatting and be parameterized by both precision and colon-vs.-empty-string separator).

That said, I can understand separating such work out from this PR. Maybe it should precede it, though?

justingrant replied:
I agree, it's a good idea for those three AOs to be coalesced editorially into something cleaner. There's also a \*lot\* of normative PRs waiting to get merged, and it's been pretty hard to coordinate all the PRs and Test262 changes.

So instead of adding another editorial roadblock before merging this one, I'd prefer to merge this PR first. I'll volunteer to build an editorial PR to simplify hh:mm[:ss[.fff…]]  formatting after all the offset-related PR dust has settled.

Sound OK?

gibson042 replied:
Yep, works for me.

justingrant commented:
> I think it's fine to make those minor editorial changes in the coming days. Let me know when everyone is happy about it, and I'll merge [tc39/test262#3862](https://github.com/tc39/test262/pull/3862) and follow it up by pushing a commit to update test262 and then merge this.

@ptomato I think this one is ready to merge. After it's merged I can rebase #2574 on the updated main, and then that PR should be ready to go as well after tests are reviewed and merged.
